datasets/locations_directions.py

```diff
@@ -1,28 +1,116 @@
+import re
+from json import JSONDecodeError
 from dataclasses import dataclass
-from random import choice
 from typing import List, Tuple
-
+from utils.openai import make_user_message
+from utils.json_utils import LLMJSONError
+from goodai.helpers.json_helper import sanitize_and_parse_json
 from datasets.locations import (
     LocationsDataset,
+    LOCATIONS,
     DISTANCES,
     DIRECTIONS,
 )
 
+DirectionDict = dict[str, str | int]
+
 
 @dataclass
 class LocationsDirectionsDataset(LocationsDataset):
     name: str = "Locations Directions"
-    question: str = "Given the points of interest that I have told you about, how would I travel from {{origin}} to {{place}} following those interesting points?"
+    question: str = (
+        "Given the points of interest that I have told you about, how would I travel from {{origin}} to {{place}} "
+        "following those interesting points?"
+    )
 
     def generate_answer(self, location_info: List[Tuple[str, str, str, int]]) -> Tuple[Tuple[str, int], List[str]]:
         # Choose the direction, and distance - it doesn't really matter
-        last_move = (choice(DIRECTIONS), choice(DISTANCES))
+        last_move = (self.random.choice(DIRECTIONS), self.random.choice(DISTANCES))
 
         # Now the answer should include all the directions and locations
-        answer = ""
-        for l in location_info[1:-1]:
-            answer += f"From {l[1]}, go {l[3]}km {l[2]} to {l[0]}.\n"
+        directions = ["(Just an example, if exact instructions are followed)"]
+        for destination, origin, direction, distance in location_info[1:-1]:
+            directions.append(f"From {origin}, go {distance}km {direction} to {destination}.")
 
-        answer += f"From {location_info[-1][1]}, go {last_move[1]}km {last_move[0]} to {location_info[-1][0]}."
+        directions.append(f"From {location_info[-1][1]}, go {last_move[1]}km {last_move[0]} to {location_info[-1][0]}.")
+        answer = "\n".join(directions)
 
         return last_move, [answer]
+
+    def parse_directions(self, expected_answer: str) -> list[DirectionDict]:
+        pattern = r"^From (?P<origin>\w+(?: \w+){0,2}), go (?P<distance>\d)km (?P<direction>\w+) to (?P<destination>\w+(?: \w+){0,2}).$"
```

Comment:
AAAAAAAAAAAAAAAAAAAAAAAAAAAAA

Reply:
:joy: Some hints:
- If it's between brackets, it will extract it into a group.
- `(?P<name>...)` Means that the group will have the key `name` (I looked it up in stackoverflow).
- `(?:...)` Means that it shouldn't extract this i.e. the brackets are just for grouping it in the expression.